langpuppet/merge.go

```diff
@@ -0,0 +1,135 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package langpuppet implements an integration entrypoint that combines lang and Puppet.
+package langpuppet
+
+import (
+	"fmt"
+	"github.com/purpleidea/mgmt/pgraph"
+	"github.com/purpleidea/mgmt/engine"
+	"strings"
+)
+
+// This method returns the merged graph containing all vertices
+// and edges found in the graphs produced by the lang and Puppet GAPIs
+// associated with the wrapping GAPI.
+// Vertices are merged if they adhere to the following rules (for any
+// given value of POSTFIX):
+//  1. The graph from lang contains a noop vertex named puppet_POSTFIX
+//  2. The graph from Puppet contains an empty class mgmt_POSTFIX
+//  3. The resulting graph will contain one noop vertex named POSTFIX
+//     that replaces all nodes mentioned in (1) and (2)
+// All edges connecting to any of the vertices merged this way
+// will be present in the merged graph.
+func (obj *GAPI) mergeChildGraphs() (*pgraph.Graph, error) {
```

Comment:
I'll leave all this to you. =D Unless you have questions. However... If you don't plan on using any of the GAPI stuff in here, I'd instead implement this as a simple merge disconnected to obj.

Eg:

```
func (g1, g2 *pgraph.Graph) (*pgraph.Graph, error)
```

And in the rare chance that it makes sense, you could always do:
```
func (graphs ...*pgraph.Graph) (*pgraph.Graph, error)
```

;)

Reply:
I took my cue from the puppet stuff, where the functions in puppet.go are actual methods, although they probably don't need to be. (Didn't look into this too deeply, though, and agree that it's not required here at all.)

I *would* be grateful for feedback regarding my use of the Pgraph API, and if my handling of vertices and edges is sound. Look for variables `result`, `lg`, and `pg`.

Reply:
Made the merge func stand alone.